compiler/rustc_infer/src/infer/error_reporting/suggest.rs

```diff
@@ -461,6 +461,62 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
             }
         }
     }
+
+    /// For "one type is more general than the other" errors on closures, suggest changing the lifetime
+    /// of the parameters to accept all lifetimes.
+    pub(super) fn suggest_for_all_lifetime_closure(
+        &self,
+        span: Span,
+        exp_found: &ty::error::ExpectedFound<ty::PolyTraitRef<'tcx>>,
+        diag: &mut Diagnostic,
+    ) {
+        // 1. Get the substs of the closure.
+        // 2. Assume exp_found is FnOnce / FnMut / Fn, we can extract function parameters from [1].
+        let expected = exp_found.expected.map_bound(|x| x.substs.get(1).cloned()).transpose();
+        let found = exp_found.found.map_bound(|x| x.substs.get(1).cloned()).transpose();
+
+        // 3. Extract the tuple type from Fn trait and suggest the change.
+        if let (Some(expected), Some(found)) = (expected, found) {
+            let expected = expected.skip_binder().unpack();
+            let found = found.skip_binder().unpack();
+            if let (GenericArgKind::Type(expected), GenericArgKind::Type(found)) = (expected, found)
```

Comment:
If you're gonna use let-chains here, can you split this `let (a, b) = (c, d)` into `let a = c && let b = d`? It's confusing to mix these.